structure_tests/drivers/dockerDriver.go

```diff
@@ -0,0 +1,298 @@
+	d.OriginalImage = newImage
+	d.CurrentImage = d.OriginalImage
+	return nil
+}
+
+func (d DockerDriver) ResetEnvVars(t *testing.T, vars []unversioned.EnvVar) {
+	// since the container will be destroyed after the tests, this is a noop
+}
+
+// copies a file from a docker container to the local fs, and returns its path
+// caller is responsible for removing this file when finished
+func (d DockerDriver) retrieveFile(t *testing.T, path string, directory bool) (string, error) {
+	ctx := context.Background()
+
+	// TODO(nkubala): this contains a hack to get around the fact that
+	// the client will not allow creating a container without a command.
+	// we should remove this, or at the very least find a better alternative
+	// given that not every container is guaranteed to have a shell.
+	container, err := d.cli.CreateContainer(docker.CreateContainerOptions{
+		Config: &docker.Config{
+			Image:        d.CurrentImage,
+			Cmd:          []string{"/bin/sh"},
+			AttachStdout: true,
+			AttachStderr: true,
+		},
+		HostConfig:       nil,
+		NetworkingConfig: nil,
+		Context:          ctx,
+	})
+	if err != nil {
+		t.Errorf("Error creating container: %s", err.Error())
+		return "", err
+	}
+
+	if directory {
+		return "", nil
+	} else {
+		tmpFile, err := ioutil.TempFile("", "structure_test")
+		if err != nil {
+			t.Errorf("Error when creating temp file: %s", err.Error())
+			return "", err
+		}
+		stream := bufio.NewWriter(tmpFile)
+
+		err = d.cli.DownloadFromContainer(container.ID, docker.DownloadFromContainerOptions{
+			OutputStream: stream,
+			Path:         path,
+			Context:      ctx,
+		})
+		if err != nil {
+			t.Errorf("Error when downloading file from container: %s", err.Error())
+			return "", err
+		}
+		stream.Flush()
+		tmpFile.Close()
+		return tmpFile.Name(), nil
+	}
+}
+
+func (d DockerDriver) StatFile(t *testing.T, path string) (os.FileInfo, error) {
+	file, err := d.retrieveFile(t, path, false)
+	if err != nil {
+		return nil, err
+	}
+	defer os.Remove(file)
+
+	f, err := os.Stat(file)
+	if err != nil {
+		return nil, err
+	}
+	if f == nil {
+		return nil, nil
+	}
+	return f, nil
+}
+
+func (d DockerDriver) ReadFile(t *testing.T, path string) ([]byte, error) {
+	file, err := d.retrieveFile(t, path, false)
+	if err != nil {
+		return nil, err
+	}
+	defer os.Remove(file)
+	return ioutil.ReadFile(file)
+}
+
+func (d DockerDriver) ReadDir(t *testing.T, path string) ([]os.FileInfo, error) {
+	// TODO(nkubala): unimplemented
+	return nil, nil
+}
+
+// This method takes a command (in the form of a list of args), and does the following:
+// 1) creates a container, based on the "current latest" image, with the command set as
+// the command to run when the container starts
+// 2) starts the container
+// 3) commits the container with its changes to a new image,
+// and sets that image as the new "current image"
+func (d DockerDriver) runAndCommit(t *testing.T, env []string, command []string) string {
+	name := utils.GenerateContainerName()
```

Comment:
Can we get docker to generate a name for us somehow? What if you don't set one?

Reply:
yeah I was originally using this name to do some other container magic, but we don't need it anymore. I should be able to just pass the container ID instead so I'll remove this